metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/lineage.py

```diff
@@ -962,3 +964,77 @@ def test_capability(self, project_id: str) -> None:
+        for source_uri in source_uris:
+            # Check that storage_location have the gs:// prefix.
+            # Right now we are only supporting GCS lineage
+            if not gcs_utils.is_gcs_uri(source_uri):
+                continue
+            regex = self.pattern_to_regex_for_gcs(
+                gcs_utils.strip_gcs_prefix(source_uri)
+            )
+            matching_urns = {
+                urn for urn in urns if regex.match(self.extract_path(urn) or "")
+            }
+
+            if not matching_urns:
+                continue
+
+            upstreams_list.extend(
+                [
+                    UpstreamClass(
+                        dataset=source_dataset_urn, type=DatasetLineageTypeClass.VIEW
+                    )
+                    for source_dataset_urn in matching_urns
+                ]
+            )
+
+        if not upstreams_list:
+            return
+
+        upstream_lineage = UpstreamLineageClass(upstreams=upstreams_list)
+        yield from self.gen_lineage(dataset_urn, upstream_lineage)
+
+    def extract_path(self, urn: str) -> Optional[str]:
+        parts = urn.split(",")
+        if len(parts) > 1:
+            return parts[1]
+        return None
+
+    def pattern_to_regex_for_gcs(self, pattern: str) -> re.Pattern:
+        """
+        supported pattern to get source_dataset_urns for gcs
+        gs://my-bucket/foo/tests/bar.avro # single file table
+        gs://my-bucket/foo/tests/*.* # mulitple file level tables
+        gs://my-bucket/foo/tests/{table}/*.avro #table without partition
+        gs://my-bucket/foo/tests/{table}/*/*.avro #table where partitions are not specified
+        gs://my-bucket/foo/tests/{table}/*.* # table where no partitions as well as data type specified
+        gs://my-bucket/{dept}/tests/{table}/*.avro # specifying keywords to be used in display name
+        gs://my-bucket/{dept}/tests/{table}/{partition_key[0]}={partition[0]}/{partition_key[1]}={partition[1]}/*.avro # specify partition key and value format
+        gs://my-bucket/{dept}/tests/{table}/{partition[0]}/{partition[1]}/{partition[2]}/*.avro # specify partition value only format
+        gs://my-bucket/{dept}/tests/{table}/{partition[0]}/{partition[1]}/{partition[2]}/*.* # for all extensions
+        gs://my-bucket/*/{table}/{partition[0]}/{partition[1]}/{partition[2]}/*.* # table is present at 2 levels down in bucket
+        gs://my-bucket/*/*/{table}/{partition[0]}/{partition[1]}/{partition[2]}/*.* # table is present at 3 levels down in bucket
+        """
+        pattern = re.escape(pattern)
+        pattern = re.sub(
+            r"\\\{[^}]+\\\}", "[^/]+", pattern
+        )  # Replace anything inside {} with [^/]+
+        pattern = pattern.replace(
+            r"\*", "[^/]*"
+        )  # Replace * with [^/]* to match filenames
+        pattern = pattern.replace(
+            r"\.\*", r"\.[^/]+"
+        )  # Adjust to match file extensions
+        return re.compile(f"^{pattern}$")
```

Comment:
this does not seem to be implementing path spec correctly

Reply:
Hi @hsheth2,

We're working on handling GCS (Google Cloud Storage) paths in a manner similar to S3 paths for BigQuery external tables. Specifically, we use the `pattern_to_regex_for_gcs` function to convert a GCS path pattern into a regex pattern. This allows us to accurately match files based on their GCS path format.

For example, pattern `sample_bigquery_data/bq_tables/*.parquet`, it generates a regex pattern that matches any .parquet file in the bq_tables directory within sample_bigquery_data, ensuring precise matching from start to end of the string.

in above case we will get
sample_bigquery_data/bq_tables/test1.parquet
sample_bigquery_data/bq_tables/test2.parquet

Once we have the complete GCS URN, we map it to the corresponding BigQuery table for lineage tracking.

Do you have any other suggestions or alternative solutions if my understanding is correct?